torch/csrc/jit/codegen/cuda/lower_validation.cpp

```diff
@@ -605,6 +606,165 @@ void validateParallelize(Fusion* fusion) {
   }
 }
 
+namespace {
+
+// Backward propagation of partial ranges from outputs to
+// inputs. Necessary to determine required ranges to compute.
+//
+// Example:
+//  tv0: [0:N]
+//  tv1: shift(tv0, {1}) -> [1:N]
+//  tv2: shift(tv0, {-1}) -> [0:N-1]
+//  tv3: tv1 + tv2 -> [1:N-1]
+//
+// In this case, the valid range of tv3 starts at 1 and ends at
+// N-1. This means that not all of the values of tv1 and tv2 are
+// actually necessary. Specifically, tv1[0] and tv2[N-1] aren't used
+// for tv3. This function calculates the required minimum range of
+// each tensor that needs to be computed.
+std::unordered_map<IterDomain*, std::pair<int64_t, int64_t>> getLiveRangeOffsets(
+    Fusion* fusion) {
+  auto exprs = ExprSort::getExprs(fusion);
+
+  std::unordered_map<IterDomain*, std::pair<int64_t, int64_t>> map;
+
+  ExpressionEvaluator ee(fusion);
+
+  for (auto it = exprs.rbegin(); it != exprs.rend(); ++it) {
+    auto expr = *it;
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      for (auto consumer_root : consumer->getRootDomain()) {
+        auto consumer_start = ee.evaluate(consumer_root->start());
+        auto consumer_stop = ee.evaluate(consumer_root->stopOffset());
+        TORCH_INTERNAL_ASSERT(
+            consumer_start.has_value(),
+            "Can't evaluate start value of ",
+            consumer_root);
+        TORCH_INTERNAL_ASSERT(
+            consumer_stop.has_value(),
+            "Can't evaluate stop value of ",
+            consumer_root);
+        auto& consumer_range =
+            map.insert({consumer_root, {0, 0}}).first->second;
+        consumer_range.first =
+            std::max(consumer_range.first, consumer_start.value());
+        consumer_range.second =
+            std::max(consumer_range.second, consumer_stop.value());
+      }
+      for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+        auto c2p =
+            PairwiseRootDomainMap(producer, consumer)
+                .mapConsumerToProducer(consumer->domain(), producer->domain());
+        for (auto consumer_root : consumer->getRootDomain()) {
+          auto producer_it = c2p.find(consumer_root);
+          if (producer_it == c2p.end()) {
+            continue;
+          }
+          auto producer_root = producer_it->second;
+          auto& consumer_range = map.at(consumer_root);
+          auto& producer_range =
+              map.insert({producer_root, {0, 0}}).first->second;
+          producer_range.first =
+              std::max(producer_range.first, consumer_range.first);
+          producer_range.second =
+              std::max(producer_range.second, consumer_range.second);
+        }
+      }
+    }
+  }
+
+  return map;
+}
+
+// Condition to check:
+// split_offset <= id_offset + halo
```

Comment:
That's a good point. I implicitly assumed both start and stop offsets are small integers, but technically it's possible that non-padded shift generates an empty domain. 

I believe it's still fine since we check the start and stop predicates separately. When `start > stop`, no computation is done because a loop index must be `>= start` and `< stop`. If that should be flagged as an error, then we would need a runtime check.